### PR TITLE
Show and filter dotty versions

### PR DESCRIPTION
**Pull request by adpi2**
You can see the result here https://index-dev.scala-lang.org

### CODE REVIEW

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
+package ch.epfl.scala.index.model
+package release
+
+import fastparse.NoWhitespace._
+import fastparse._
+
+sealed trait BackendVersion {
```
Comment (julienrf) [inline model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:7]: It would be nice to have some scaladocs explaining what this sealed trait models, and what `family` and `isValid` are.

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      `0.19`.version,
+      `0.20`.version,
+      `0.21`.version,
+      `0.22`.version,
+      `0.23`.version
+    )
+
+  def isValid(version: BinaryVersion): Boolean =
+    latestVersions.contains(version) ||
+      (version > `0.23`.version && version < MinorBinary(0, 30))
```
Comment (julienrf) [inline model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:71]: Why not just this?

```suggestion
  def isValid(version: BinaryVersion): Boolean = version > `0.16`.version
```

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala (hunk 1 above)
Comment (julienrf) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:7]: Also, I’m not sure `BackendVersion` is a good name for that. I think “binary version” or “scala version” would be better.

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala (hunk 2 above)
Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:71]: The reasons are:
- we want to distiguinsh between scala version and dotty version: `2.12` is not a valid dotty version
- previous milestones and RC are not relevant: `0.22-RC1` is not a valid dotty version
but a future `0.24-RC1` is a valid dotty version
- some sbt plugin have bad version: `_0.31_2.12` instead of `_2.12_0.31` and we don't want to interpret this as a plugin for sbt 2.12 and dotty 0.31

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala (hunk 1 above)
Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:7]: Agreed for `ScalaVersion` and for the scaladocs

Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:7]: Ok for the scaladoc

Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:7]: I dont think that `BinaryVersion` is a better name because there already is a `BinaryVersion` ADT that is used for describing the scala-js version and the scala-native version

Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:7]: And the problem with `ScalaVersion` is to find a name of the two subtypes that are currently named `ScalaBackend` and `DottyBackend`. Is 0.23 a ScalaVersion? My answer would be not yet because dotty is not Scala 3 yet. But still dotty can be used as a backend for the javascript and scala-native compilers. That is why I choosed the name `BackendVersion`. A scala js target is made of a `BackendVersion` and a javascript `BinaryVersion`. It makes sense to me.

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala (hunk 2 above)
Comment (julienrf) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:71]: I see, thanks for the explanation!

Does that mean that we will have to update scaladex each time a new Dotty RC is released?

Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:71]: That's a good question. As of now I have seen no libraries issued with a Dotty RC version. But we will have to update Scaladex when Dotty version 0.30 will be released. And this is also true for scala 3.0 scala.js 2.x and scala-native 0.4 (not sure for this one) and sbt 2.x.

Hopefully we will make the ScalaTarget parser more robust and then we will make it forward compatible.

Comment (julienrf) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:71]: Maybe we can assume for now that Dotty libraries will always target non-RC Dotty releases? So maybe we can write a more general test that checks that the major version is 0 and that the patch version is a number (it doesn’t contain “RC...”). Wdyt?

Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:71]: Yes it would be better. But we still want to discard the versions older than 0.17 and the badly formated sbt versions.

Comment (julienrf) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:71]: For badly formatted sbt versions, I believe we could make a more robust parser: it should only accept one binary suffix, not two.

Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala]: I don't really know where do those sbt target versions come frome but they have two parts, a scala version and a sbt version. Most of the time we have _<scala version>_<sbt version> but sometimes they are inverted.

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala (hunk 1 above)
Comment (sjrd) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:11]: *Backend* is definitely the wrong word. If you want to reason along your reasoning, it should be *frontend* version. Scala 2/Dotty are frontends, not backends. Backends would be JVM/JS/Native. But even there for JS itself we have a front-end and a back-end. It's all very relative. But in any case the distinction scalac/dotty is the very first "end" of the language, so in any naming scheme it would be a frontend, not a backend.

But I think the best name for this---independently of the philosophical debate whether Dotty is "Scala" or not---is `LanguageVersion`. This definitely represents a version of the *language*, whether that language is called "Scala" or "Dotty". So I would recommend `LanguageVersion` with the subtypes `ScalaVersion` and `DottyVersion`.

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+  def Parser[_: P]: P[ScalaBackend] = {
+    BinaryVersion.Parser
+      .filter(isValid)
+      .map(ScalaBackend(_))
+  }
+}
+
+object DottyBackend {
+  val `0.23`: DottyBackend = DottyBackend(MinorBinary(0, 23))
```
Comment (sjrd) [inline model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:63]: Why singling out any Dotty version? A new one comes up every 6 weeks. You don't want to have to modify this codebase every 6 weeks.

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+  def isValid(version: BinaryVersion): Boolean =
+    version match {
+      case MinorBinary(major, minor) if major == 0 && minor < 30 => true
```
Comment (sjrd) [inline model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:67]:
```suggestion
      case MinorBinary(major, minor) if major == 0 => true
```
All 0.x versions are Dotty version. There's no artificial cap at 30.

File: model/src/main/scala/ch.epfl.scala.index.model/release/ScalaTarget.scala [hunk 5]
```diff
@@ -37,86 +41,92 @@ sealed trait ScalaTarget extends Ordered[ScalaTarget] {
...
+  }
 
-  def isValid(version: BinaryVersion): Boolean =
-    stableBinaryVersions.contains(version)
+  def Parser[_: P]: P[ScalaTarget] =
+    ("_" ~ BackendVersion.Parser).map(ScalaJvm.apply)
 }
 
-object ScalaJs {
+object ScalaJs extends Parsers {
```
Comment (sjrd) [inline model/src/main/scala/ch.epfl.scala.index.model/release/ScalaTarget.scala:100]: Why is this necessary? `ScalaNative` below doesn't have it.

Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/ScalaTarget.scala:100]: it is not :)

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala (hunk 2 above)
Comment (julienrf) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala]: The thing that I don’t understand is why is a pattern such as `_x.y_α.β` successfully parsed in the first place as a valid Scala version?

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala (hunk 1 above)
Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:11]: You are totally right. I have inverted the meaning of Backend and Frontend in my head. `LangugageVersion` as you suggest is a better name.

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala (hunk 3 above)
Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:63]: This is just a constant value that is used in the tests. But we dont really need it.

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala (hunk 4 above)
Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:67]: Right but I definetely dont want to interpret this sbt plugin version _0.31_2.12 version as a plugin of sbt  2.12 compiled with Dotty 0.31. Would a `isValid` sbt version that discard sbt `2.12` be better?

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala (hunk 2 above)
Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala]: Its not a Scala version its a Sbt target version. Nevertheless I've moved the logic of discarding bad sbt plugin target to SbtPlugin which is indeed better.

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala (hunk 4 above)
Comment (adpi2) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala:67]: Yes I think it is actually

File: model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala (hunk 3 above)
Comment (sjrd) [reply model/src/main/scala/ch.epfl.scala.index.model/release/BackendVersion.scala]: If it is only relevant in the tests, you should declare it in the tests ;)
